packages/browser/src/core/page/add-page-context.ts

```diff
@@ -0,0 +1,33 @@
+import { pick } from '../../lib/pick'
+import { EventProperties, SegmentEvent } from '../events'
+import { getDefaultPageContext } from './get-page-context'
+
+/**
+ * Augments a segment event with information about the current page.
+ * Page information like URL changes frequently, so this is meant to be captured as close to the event call as possible.
+ * Things like `userAgent` do not change, so they can be added later in the flow.
+ * We prefer not to add this information to this function, as it increases the main bundle size.
+ */
+export const addPageContext = (
+  event: SegmentEvent,
+  pageCtx = getDefaultPageContext()
+): void => {
+  const evtCtx = event.context! // Context should be set earlier in the flow
+  let pageContextFromEventProps: Pick<EventProperties, string> | undefined
+  if (event.type === 'page') {
+    pageContextFromEventProps =
+      event.properties && pick(event.properties, Object.keys(pageCtx))
+
+    event.properties = {
+      ...pageCtx,
+      ...event.properties,
+      ...(event.name ? { name: event.name } : {}),
+    }
+  }
+
+  evtCtx.page = {
+    ...pageCtx,
+    ...pageContextFromEventProps,
+    ...evtCtx.page,
+  }
```

Comment:
In general I prefer to keep these kind of functions immutable so that there's no invisible sideeffects

Reply:
I more or less copied and pasted this from the plugin, which is why it mutates. You mean have `addPageContext` return a new event, rather than mutate?

Reply:
I believe the answer is "it does not", but these mutations aren't modifying the properties object the customer passes in right?

Reply:
> These mutations aren't modifying the properties object the customer passes in right?
Correct